File: src/main/java/CodeGenerator.java
import annotations.BindMethod;

import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.Arrays;
import java.util.Comparator;
import java.util.stream.Collectors;

public final class CodeGenerator {
    private final static String CLIENT_FOLDER_PATH = "frontend/javatron/";
    private final static String METHODS_FOLDER_PATH = CLIENT_FOLDER_PATH + "methods/";
    private final static String TYPES_FILE_PATH = CLIENT_FOLDER_PATH + "types.ts";

    static {
        FileManager.createNewFile(TYPES_FILE_PATH);
        FileManager.createOrReplaceDirectory(METHODS_FOLDER_PATH);
    }

    public static void generateTypes(Object... objects) {
        FileWriter writer = null;
        try {
            Class<?>[] classes = TypeConverter.getClasses(objects);
            writer = new FileWriter(TYPES_FILE_PATH, true);
            Arrays.stream(classes).forEach(c -> TypeConverter.boundTypes.add(c.getSimpleName()));

            for (Class<?> clazz : classes) {
                // Declare type and export
                String typeName = clazz.getSimpleName();
                writer.write("export type " + clazz.getSimpleName() + " = {\n");

                // Add fields and map the types from java to typescript
                Field[] fields = clazz.getDeclaredFields();
                for (Field field : fields) {
                    String name = field.getName();
                    String type = TypeConverter.convert(field.getGenericType(), false);
                    writer.write("\t" + name + ": " + type + "\n");
                }
                writer.write("}\n\n");

                System.out.println("Created type: " + typeName);
            }
        } catch (IOException e) {
            System.out.println("Failed to create types for frontend.");
            e.printStackTrace();
            System.exit(1);
        } finally {
            try {
                if (writer != null)
                    writer.close();
            } catch (IOException closeErr) {
                System.out.println("Failed to close writer.");
                closeErr.printStackTrace();
            }
        }
    }

    public static void generateFunctions(Object... objects) {
        Class<?>[] classes = TypeConverter.getClasses(objects);
        for (Class<?> c : classes) {
            createJavascriptFunctions(c);
            createTypescriptDeclarations(c);
        }
    }

    private static void createJavascriptFunctions(Class<?> c) {
        FileWriter writer = null;
        try {
            String className = c.getSimpleName();
            String path = METHODS_FOLDER_PATH + className + ".js";
            FileManager.createNewFile(path);
            writer = new FileWriter(path);

            for (Method method : c.getDeclaredMethods()) {
                if (!method.isAnnotationPresent(BindMethod.class)) continue;

                String methodName = method.getName();
                Parameter[] params = method.getParameters();
                String argsString = Arrays.stream(params).map(Parameter::getName).collect(Collectors.joining(","));
                writer.write("export function " + methodName + "(" + argsString + ") {\n");
                writer.write("\treturn window[\"" + className + "_" + methodName + "\"](" + argsString + ");\n");
                writer.write("}\n\n");
            }
        } catch (IOException e) {
            System.out.println("Failed to create javascript function for class " + c.getSimpleName() + ".");
            e.printStackTrace();
        } finally {
            try {
                if (writer != null) {
                    writer.close();
                }
            } catch (IOException closeErr) {
                System.out.println("Failed to close writer");
                closeErr.printStackTrace();
            }
        }
    }

    private static String toJSType(Class<?> c) {
        String type = TypeConverter.jsTypes.get(c.getSimpleName());
        if (type != null) return type;

        return "jt." + c.getSimpleName();
    }

    private static void createTypescriptDeclarations(Class<?> c) {
        FileWriter writer = null;
        try {
            String className = c.getSimpleName();
            String path = METHODS_FOLDER_PATH + className + ".d.ts";
            FileManager.createNewFile(path);
            writer = new FileWriter(path);

            writer.write("import * as jt from '../types';\n\n");

            for (Method method : Arrays.stream(c.getDeclaredMethods()).sorted(Comparator.comparing(Method::getName)).toList()) {
                if (!method.isAnnotationPresent(BindMethod.class)) continue;
                System.out.println(method.getName());
                String methodName = method.getName();
                Parameter[] params = method.getParameters();
                String argsString = Arrays.stream(params)
                        .map(p -> {
                            String name = p.getName();
                            String type = toJSType(p.getType());
                            System.out.println("type: " + type);
                            return name + ": " + type;
                        })
                        .collect(Collectors.joining(","));
                String convertedReturnType = TypeConverter.convert(method.getReturnType(), true);
                String returnTypeString = convertedReturnType.equals("void") ? "" : ": Promise<" + convertedReturnType + ">";
                writer.write("export function " + methodName + "(" + argsString + ")" + returnTypeString + ";\n\n");
                System.out.println("----------");
            }
        } catch (IOException e) {
            System.out.println("Failed to create typescript declarations for class " + c.getSimpleName() + ".");
            e.printStackTrace();
        } finally {
            try {
                if (writer != null) {
                    writer.close();
                }
            } catch (IOException closeErr) {
                System.out.println("Failed to close writer.");
                closeErr.printStackTrace();
            }
        }
    }
}

File: src/main/java/CodeGenerator.kt
import annotations.BindMethod
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.io.PrintWriter
import kotlin.system.exitProcess

object CodeGenerator {
    private const val CLIENT_FOLDER_PATH = "frontend/javatron/"
    private const val METHODS_FOLDER_PATH = CLIENT_FOLDER_PATH + "methods/"
    private const val TYPES_FILE_PATH = CLIENT_FOLDER_PATH + "types.ts"

    private const val EVENTS_API_FILE_RESOURCE = "events.ts"
    private const val WINDOW_DECLARE_FILE_RESOURCE = "window.d.ts"

    private const val EVENTS_API_FILE_DEST = CLIENT_FOLDER_PATH + "events.ts"
    private const val WINDOW_DECLARE_FILE_DEST = CLIENT_FOLDER_PATH + "window.d.ts"

    private val log = LoggerFactory.getLogger(this::class.java)

    init {
        FileManager.createOrReplaceFile(TYPES_FILE_PATH)
        FileManager.createOrReplaceDirectory(METHODS_FOLDER_PATH)
    }

    fun generateEventsAPI() {
        try {
            FileManager.createOrReplaceFile(EVENTS_API_FILE_DEST)
            FileManager.createOrReplaceFile(WINDOW_DECLARE_FILE_DEST)

            val eventsFileURL = object {}.javaClass.classLoader.getResource(EVENTS_API_FILE_RESOURCE)
            val windowFileURL = object {}.javaClass.classLoader.getResource(WINDOW_DECLARE_FILE_RESOURCE)

            if (eventsFileURL != null && windowFileURL != null) {
                val eventsFileContents = File(eventsFileURL.toURI()).readText()
                val windowFileContents = File(windowFileURL.toURI()).readText()

                File(EVENTS_API_FILE_DEST).printWriter().use { out -> out.println(eventsFileContents) }
                File(WINDOW_DECLARE_FILE_DEST).printWriter().use { out -> out.println(windowFileContents) }
            } else {
                throw Exception("Failed to find files in resources.")
            }
        } catch (e: Exception) {
            log.error("Error reading the file: ${e.message}")
            exitProcess(1)
        }
        log.info("Created events API files.")
    }

    fun generateTypes(vararg objects: Any) {
        try {
            val writer = PrintWriter(TYPES_FILE_PATH)
            val classes = objects.map { it.javaClass }
            TypeConverter.boundTypes.addAll(classes.map { it.simpleName })
            for (c in classes) {
                // Declare type and export
                writer.println("export type ${c.simpleName} = {")

                // Add fields and map the types from java to typescript
                for (field in c.declaredFields) {
                    val name = field.name
                    val type = TypeConverter.convert(field.genericType, false)
                    writer.println("\t$name: $type")
                }
                writer.println("}\n")

                log.info("Created type: ${c.simpleName}")
            }
            writer.close()
        } catch (e: IOException) {
            log.error("Failed to generate a types file `types.ts`.", e)
            exitProcess(1)
        }
    }

    fun generateFunctions(vararg objects: Any) {
        for (c in objects.map { it.javaClass }) {
            createJavascriptFunctions(c)
            createTypescriptDeclarations(c)
        }
    }

    private fun createJavascriptFunctions(c: Class<*>) {
        try {
            val className = c.simpleName
            val path = "$METHODS_FOLDER_PATH$className.js"
            FileManager.createOrReplaceFile(path)
            val writer = PrintWriter(path)

            for (method in c.declaredMethods) {
                if (!method.isAnnotationPresent(BindMethod::class.java)) continue

                val methodName = method.name
                val argsString = method.parameters.joinToString(",") { it.name }
                writer.println("export function $methodName($argsString) {")
                writer.println("\treturn window[\"${className}_$methodName\"]($argsString);")
                writer.println("}\n")
            }

            writer.close()
        } catch (e: IOException) {
            log.error("Failed to create javascript function for class ${c.simpleName}.", e)
            exitProcess(1)
        }
    }

    private fun createTypescriptDeclarations(c: Class<*>) {
        try {
            val path = "$METHODS_FOLDER_PATH${c.simpleName}.d.ts"
            FileManager.createOrReplaceFile(path)
            val writer = PrintWriter(path)

            writer.println("import * as t from '../types';\n")

            for (method in c.declaredMethods.sortedBy { it.name }) {
                if (!method.isAnnotationPresent(BindMethod::class.java)) {
                    continue
                }

                val argsString = method.parameters.joinToString(", ") {
                    "${it.name}: ${TypeConverter.convert(it.parameterizedType, true)}"
                }
                val convertedReturnType = TypeConverter.convert(method.genericReturnType, true)
                var returnTypeString = ": Promise<$convertedReturnType>"
                if (convertedReturnType == "void") returnTypeString = ""

                writer.println("export function ${method.name}($argsString)$returnTypeString;\n")

                log.info("Created function ${method.name} of class ${c.simpleName}")
            }

            writer.close()
        } catch (e: IOException) {
            log.error("Failed to create typescript declarations for class ${c.simpleName}.", e)
            exitProcess(1)
        }
    }
}